commons/zenoh-buffers/src/bbuf.rs

```diff
@@ -127,15 +127,15 @@ impl Writer for &mut BBuf {
         self.capacity() - self.len()
     }
 
-    fn with_slot<F>(&mut self, len: usize, f: F) -> Result<NonZeroUsize, DidntWrite>
+    unsafe fn with_slot<F>(&mut self, len: usize, write: F) -> Result<NonZeroUsize, DidntWrite>
     where
         F: FnOnce(&mut [u8]) -> usize,
     {
         if self.remaining() < len {
             return Err(DidntWrite);
         }
 
-        let written = f(self.as_writable_slice());
+        let written = write(self.as_writable_slice());
```

Comment:
I think `as_writable_slice` should be unsafe as well. It claims that the `Writer` ensures that `self.len` is less than the capacity.

The invariant was proved in `with_slot` before, and now it's the responsibility of the caller of `with_slot`. `unsafe` should not be used unless the current context can prove there is no UB. `as_writable_slice` cannot be safe if its invariant can be broken w/ safe code.

Reply:
The same applies to `as_mut_slice`.

Reply:
I think you're wrong there. It's the responsability of the caller of `with_slot`, but because `with_slot` is `unsafe`, it's fine.

There is nothing wrong to have the possibility of having invariant broken by *unsafe* code. For example, you can break  `Vec` invariant using `set_len`, and every safe methods of `Vec` will be corrupted; as `set_len` is `unsafe`, it's fine. 
It's the same here: if `self.len` is updated without checking the invariant, it must be done in `unsafe` method.